ui/lobby/src/setupCtrl.ts

```diff
@@ -0,0 +1,282 @@
+  ratingRange = (): string => {
+    if (!this.root.data.ratingMap) return '';
+    const rating = this.root.data.ratingMap[this.selectedPerf()];
+    return `${rating + this.ratingMin()}-${rating + this.ratingMax()}`;
+  };
+
+  // This function is a special version of the above function that does not require the modal to
+  // be opened. Used in boot.ts for hook-like games ("New opponent" button).
+  hookRatingRange = (): string => {
+    if (!this.root.data.ratingMap) return '';
+    const { variant, timeMode, time, increment, ratingMin, ratingMax } = this.makeSetupStore('hook')();
+    const rating = this.root.data.ratingMap[getPerf(variant, timeMode, time, increment)];
+    return `${rating + ratingMin}-${rating + ratingMax}`;
+  };
+
+  hookToPoolMember = (color: Color | 'random'): PoolMember | null => {
+    const valid =
+      color == 'random' && this.variant() == 'standard' && this.gameMode() == 'rated' && this.timeMode() == 'realTime';
+    const id = `${this.time()}+${this.increment()}`;
+    return valid && this.root.pools.find(p => p.id === id)
+      ? {
+          id,
+          range: this.ratingRange(),
+        }
+      : null;
+  };
+
+  propsToFormData = (color: Color | 'random'): FormData =>
+    xhr.form({
+      variant: keyToId(this.variant(), variants).toString(),
+      fen: this.fen(),
+      timeMode: keyToId(this.timeMode(), timeModes(this.root.trans)).toString(),
+      time: this.time().toString(),
+      time_range: this.timeV().toString(),
+      increment: this.increment().toString(),
+      increment_range: this.incrementV().toString(),
+      days: this.days().toString(),
+      days_range: this.daysV().toString(),
+      mode: this.gameMode() === 'casual' ? '0' : '1',
+      ratingRange: this.ratingRange(),
+      ratingRange_range_min: this.ratingMin().toString(),
+      ratingRange_range_max: this.ratingMax().toString(),
+      level: this.aiLevel().toString(),
+      color,
+    });
+
+  submit = async (color: Color | 'random') => {
+    const poolMember = this.hookToPoolMember(color);
+    if (poolMember) {
+      this.root.enterPool(poolMember);
+      this.closeModal();
+      return;
+    }
+
+    if (this.gameType === 'hook') this.root.setTab(this.timeMode() === 'realTime' ? 'real_time' : 'seeks');
+    this.loading = true;
+    this.root.redraw();
+
+    let urlPath = `/setup/${this.gameType}`;
+    if (this.gameType === 'hook') urlPath += `/${lichess.sri}`;
+    const urlParams = { user: this.friendUser || undefined };
+    let response;
+    try {
+      response = await xhr.textRaw(xhr.url(urlPath, urlParams), {
+        method: 'post',
+        body: this.propsToFormData(color),
+      });
+    } catch (e) {
+      this.loading = false;
+      this.root.redraw();
+      alert('Sorry, we encountered an error while creating your game. Please try again.');
+      return;
+    }
+
+    const { ok, redirected, url } = response;
+    if (!ok) {
+      document.body.innerHTML = await response.text();
```

Comment:
Not sure this is ok to do in the general `!ok` case. Was just thinking about the anonymous challenging user case